### PR TITLE
Tokenize Unicode identifiers

### PR DESCRIPTION
**Pull request by tamaroning**
Addresses #2287 #418

- [x] add function which checks XID_Start and XID_Continue to libcpp
- [x] identifiers
- [x] lifetime token

gcc/rust/ChangeLog:
```
	* lex/rust-lex.cc (is_whitespace): add all lacked codepoints valid as whitespaces
	(is_identifier_start): new function to check XID_Start and underscore
	(is_identifier_continue): new function to check XID_Continue
	(Lexer::build_token):tokenize Unicode identifiers
	(Lexer::parse_partial_string_continue):add comments
	(Lexer::parse_partial_unicode_escape):add comments
	(Lexer::parse_raw_identifier):change to use `is_identifier_scontinue`
	(Lexer::parse_identifier_or_keyword):change to use `is_identifier_continue`
	(Lexer::parse_char_or_lifetime):change to use `is_identifier_start` and `is_identifier_continue`
	(Lexer::skip_codepoint_input):do not attempt to skip input when bumping EOF
	* lex/rust-lex.h:add `current_char32` field
```
libcpp/ChangeLog:
```
	* charset.cc (check_xid_property):new function to check XID_Start and XID_Continue
	* include/cpplib.h (check_xid_property):add forward declaration of `check_xid_property`
```

### CODE REVIEW

**Comment by tamaroning** [comment]
I found this code and am not sure how we handle width of characters 
https://github.com/tamaroning/gccrs/blob/3c0f3ca4e4d8ddeecead05e109ace1917e021321/gcc/rust/lex/rust-lex.cc#L2462-L2463

File: libcpp/charset.cc [hunk 1]
```diff
@@ -1256,6 +1256,38 @@ _cpp_uname2c_uax44_lm2 (const char *name, size_t len, char *canon_name)
...
 
+/* Returns 1 if C has the XID_Start property, 2 if C has the XID_Continue properties, 3 if C has the both properties, 0 otherwise. */
+int check_xid_property (cppchar_t c) {
+  // fast path for ASCII
+  if (c < 0x80) {
+    if (('A' <= c && c <= 'Z') || ('a' <= c && c <= 'z'))
+      return 3;
+    if (('0' <= c && c <= '9') || c == '_')
+      return 2;
+  }
```
Comment (tamaroning) [inline libcpp/charset.cc:1267]: I think this function should return bit flags but how can I do it?
(i.e. XID_Start: 1, XID_Continue: 1 << 1)
Is just using `int` (or `unsigned int`) and export constants such as `XID_START` and `XID_CONTINUE` good?

Comment (tamaroning) [reply libcpp/charset.cc]: Fix to use `unsigned int` and export consants via enum

File: gcc/rust/lex/rust-lex.cc [hunk 2]
```diff
@@ -103,11 +104,17 @@ check_valid_float_dot_end (char character)
...
 {
-  return ISSPACE (character);
+  // https://doc.rust-lang.org/reference/whitespace.html
+  return character == '\t' || character == '\n' || character == '\v'
+	 || character == '\f' || character == '\r' || character == ' '
+	 || character == 0x0085	 // next line
+	 || character == 0x200e	 // left-to-right mark
+	 || character == 0x200f	 // right-to-left mark
+	 || character == 0x2028	 // line separator
+	 || character == 0x2029; // pragraph separator
```
Comment (CohenArthur) [inline gcc/rust/lex/rust-lex.cc:117]: Are all of those characters accepted by `rustc` as whitespace?

File: gcc/rust/lex/rust-lex.cc [hunk 3]
```diff
@@ -283,22 +302,22 @@ Lexer::build_token ()
...
 	{
-	  skip_input (1);
-	  current_char = peek_input ();
-	  skip_input ();
+	  skip_input (2);
 	}
 
+      current_char = peek_input ();
+      current_char32 = peek_codepoint_input ();
+      skip_codepoint_input ();
```
Comment (CohenArthur) [inline gcc/rust/lex/rust-lex.cc:319]: why are we skipping the codepoint input here but not the char?

File: gcc/rust/lex/rust-lex.cc [hunk 4]
```diff
@@ -2443,28 +2467,29 @@ Lexer::parse_char_or_lifetime (Location loc)
...
+	      current_char32 = peek_codepoint_input ();
 	      length++;
 	    }
 
 	  current_column += length;
 
 	  loc += length - 1;
 
+	  // TODO some keywords cannot be used for a lifetime label
+	  // https://doc.rust-lang.org/reference/tokens.html
```
Comment (CohenArthur) [inline gcc/rust/lex/rust-lex.cc:2491]: That's good - please open an issue for this and put the issue number in the comment :)

**Review by CohenArthur — COMMENTED** [review]
That's great work, thank you!

File: gcc/rust/lex/rust-lex.cc (hunk 3 above)
Comment (tamaroning) [reply gcc/rust/lex/rust-lex.cc:319]: If we skip one byte here, only the first byte of current utf-8 character can be skipped by the lexer, which we do not expect. 
For example, if the lexer tokenizes identifier あああ , it should skip the first utf8 character あ, not its first byte.

File: gcc/rust/lex/rust-lex.cc (hunk 2 above)
Comment (tamaroning) [reply gcc/rust/lex/rust-lex.cc:117]: Yes. All of these values are defined in the Rust ref.
You can find URL to this just before the selected lines.

File: gcc/rust/lex/rust-lex.cc (hunk 4 above)
Comment (tamaroning) [reply gcc/rust/lex/rust-lex.cc:2491]: Opend https://github.com/Rust-GCC/gccrs/issues/2306

**Comment by tamaroning** [comment]
Oops, I mistakenly made the last commit. I will undo it later.

**Comment by tamaroning** [comment]
Done

File: gcc/rust/lex/rust-lex.cc (hunk 2 above)
Comment (CohenArthur) [reply gcc/rust/lex/rust-lex.cc:117]: Oh, I missed this! Sorry! Thanks for pointing it out haha

File: gcc/rust/lex/rust-lex.cc (hunk 3 above)
Comment (CohenArthur) [reply gcc/rust/lex/rust-lex.cc:319]: I see, thank you!

File: gcc/rust/lex/rust-lex.cc [hunk 5]
```diff
@@ -2443,28 +2467,29 @@ Lexer::parse_char_or_lifetime (Location loc)
...
+	      skip_codepoint_input ();
+	      current_char32 = peek_codepoint_input ();
 	      length++;
 	    }
 
 	  current_column += length;
 
 	  loc += length - 1;
 
+	  // TODO some keywords cannot be used for a lifetime label
```
Comment (CohenArthur) [inline gcc/rust/lex/rust-lex.cc:2490]:
```suggestion
	  // TODO some keywords cannot be used for a lifetime label #2306 
```

**Review by CohenArthur — APPROVED** [review]
Very minor nit :) This looks great to me, thank you @tamaroning!

**Comment by CohenArthur** [comment]
@tamaroning I noticed that your commit does not include a commit title - please add a short description of the commit as title. It should look something like this:

```
    gccrs: we can't check the bounds involving empty placeholder types

    We use placeholders for assoicated types on traits but if we are unifying
    types against a placeholder its not possible to check the bounds as the
    placeholder does not have enough information yet at this point to determine
    if bounds will or won't be satisfied. That check will occur when associated
    types and generics are setup.

    Fixes #2036

    gcc/rust/ChangeLog:

            * typecheck/rust-unify.cc (UnifyRules::go): dont check bounds on placeholders

    gcc/testsuite/ChangeLog:

            * rust/compile/issue-2036.rs: New test.

    Signed-off-by: Philip Herron <herron.philip@googlemail.com>
```

You can see the commit has a title, which can be anything as long as it is a short description of the commit. Then, you have the commit description - but that's not always necessary. Then the Changelog, and finally the `Signed-off-by` line

**Comment by tamaroning** [comment]
@CohenArthur  Thanks for your review! I fixed comments and commit messages.

**Review by philberty — CHANGES_REQUESTED** [review]
This patch looks great though i think what will really help is to split this into two commits. So you put all the libcpp changes into one commit and the gccrs changes in another. The only reason for this is that the libcpp changes will need to be reviewed by other people in the GCC community and it will be easier to do this by separating them out.

Great work!

**Comment by tamaroning** [comment]
> This patch looks great though i think what will really help is to split this into two commits. So you put all the libcpp changes into one commit and the gccrs changes in another. The only reason for this is that the libcpp changes will need to be reviewed by other people in the GCC community and it will be easier to do this by separating them out.

Sounds good :) Done.

**Review by P-E-P — APPROVED** [review]
I've not much to say, that's good work.

File: gcc/rust/lex/rust-lex.cc [hunk 6]
```diff
@@ -116,6 +123,18 @@ is_non_decimal_int_literal_separator (char character)
   return character == 'x' || character == 'o' || character == 'b';
 }
 
+bool
+is_identifier_start (int codepoint)
```
[the remark this answers is not in the document]
Comment (tamaroning) [reply gcc/rust/lex/rust-lex.cc:127]: Yes, I think `uint32_t` is better. If we unify types for paramters of such functions, other several functions should also use the same type.
e.g. is_x_digit, is_octal_digit, etc.
https://github.com/Rust-GCC/gccrs/blob/d535c8259301687e28105105a52e013f138b619e/gcc/rust/lex/rust-lex.cc#L83-L87

**Pull request merged**